core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java

```diff
@@ -285,16 +288,19 @@ public HttpResponse serve(ServiceRequestContext ctx, HttpRequest req) throws Exc
             synchronized (healthCheckerListener) {
                 final boolean currentHealthiness = isHealthy();
                 if (isHealthy == currentHealthiness) {
+                    final Set<PendingResponse> pendingResponses = isHealthy ? pendingUnhealthyResponses
+                                                                            : pendingHealthyResponses;
                     final CompletableFuture<HttpResponse> future = new CompletableFuture<>();
                     final ScheduledFuture<Boolean> timeoutFuture = ctx.eventLoop().schedule(
                             () -> future.complete(HttpResponse.of(notModifiedHeaders)),
                             longPollingTimeoutMillis, TimeUnit.MILLISECONDS);
                     final PendingResponse pendingResponse = new PendingResponse(method, future, timeoutFuture);
-                    if (isHealthy) {
-                        pendingUnhealthyResponses.add(pendingResponse);
-                    } else {
-                        pendingHealthyResponses.add(pendingResponse);
-                    }
+                    pendingResponses.add(pendingResponse);
+                    timeoutFuture.addListener((FutureListener<Boolean>) f -> {
+                        synchronized (healthCheckerListener) {
+                            pendingResponses.remove(pendingResponse);
```

Comment:
Question: Don't we have to synchronize the following block?
https://github.com/line/armeria/blob/69acc04e620ce52445ac23d1300508f6c95815c3/core/src/main/java/com/linecorp/armeria/server/healthcheck/HealthCheckService.java#L438-L444

Reply:
This is not a problem. I misunderstood the logic. 😢